File: section35#shi-tomas/Python/Shi-Tomas.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# author：lowkeyway time:10/22/2019

import sys
import numpy as np
import cv2 as cv
import time

def shiTomasTest(imgOri):
    maxCorners = 500;
    quality_level = 0.01;
    minDistance = 0.04;
    imgGary = cv.cvtColor(imgOri, cv.COLOR_BGR2GRAY)
    corners = cv.goodFeaturesToTrack(imgGary, maxCorners, quality_level, minDistance)

    for corner in corners:
        b = np.random.randint(0, 256)
        g = np.random.randint(0, 256)
        r = np.random.randint(0, 256)

        cv.circle(imgOri, (int(corner[0][0]), int(corner[0][1])), 2, (int(b), int(g), int(r)), 2)

    cv.imshow("imgShiTomas", imgOri)

def harrisTest(imgOri):
    blockSize = 2
    apertureSize = 3
    k = 0.04

    imgGary = cv.cvtColor(imgOri, cv.COLOR_BGR2GRAY)

    imgHarris = cv.cornerHarris(imgGary, blockSize, apertureSize, k)

    imgDst = np.zeros(imgHarris.shape, dtype=np.float32)
    cv.normalize(imgHarris, imgDst, 0, 255, cv.NORM_MINMAX)
    imgDstScale = cv.convertScaleAbs(imgDst)

    for row in range(imgDstScale.shape[0]):
        for col in range(imgDstScale.shape[1]):
            if int(imgDstScale[row, col]) > 85:
                b = np.random.randint(0, 256)
                g = np.random.randint(0, 256)
                r = np.random.randint(0, 256)
                cv.circle(imgOri, (col, row), 2, (int(b), int(g), int(r)), 2)

    cv.imshow("imgHarrisDst", imgOri)

def main_func(argv):
    imgOri = cv.imread("right13.jpg")
    if imgOri is None:
        print("Cannot load this picture!")
        return False

    imgHarris = np.copy(imgOri)
    imgShiTomas = np.copy(imgOri)


    start = time.process_time()
    harrisTest(imgHarris)
    harrisTime = time.process_time()
    shiTomasTest(imgShiTomas)
    shiTomasTime = time.process_time()

    print("Harris takes {:.2f} s" .format(harrisTime - start))
    print("Harris takes {:.2f} s" .format(shiTomasTime - harrisTime))
    cv.imshow("imgOri", imgOri)

    cv.waitKey(0)

if __name__ == '__main__':
    main_func(sys.argv)

File: section35#shi-tomas/C++/Shi-Tomas.cpp
#include <iostream>
#include <string>
#include <opencv2/opencv.hpp>
#include<time.h>

using namespace std;
using namespace cv;

static bool shiTomas(Mat imgOri);
static bool harrisTest(Mat imgOri);

RNG rng(12345);

int main()
{
	Mat imgOri = imread("right13.jpg");
	if (imgOri.empty())
	{
		cout << "Cannot load this picture!" << endl;
		return false;
	}
	clock_t start, harrisTime, shiTomasTime;
	double totaltime;
	Mat imgHarris = imgOri.clone();
	Mat imgShiTomas = imgOri.clone();

	start = clock();
	harrisTest(imgHarris);
	harrisTime = clock();

	shiTomas(imgShiTomas);
	shiTomasTime = clock();

	imshow("imgOri", imgOri);

	cout << "Harris takes [" << (double)(harrisTime - start) / CLOCKS_PER_SEC << "] s" << endl;
	cout << "shiTomas takes [" << (double)(shiTomasTime - harrisTime) / CLOCKS_PER_SEC << "] s" << endl;
	
	waitKey(0);
	return true;
}

static bool harrisTest(Mat imgOri)
{
	int blockSize = 2;
	int apertureSize = 3;
	double k = 0.04;


	Mat imgGray, imgHarris;
	//showImgPara(imgGray);
	cvtColor(imgOri, imgGray, COLOR_BGR2GRAY);

	imgHarris = Mat::zeros(imgOri.size(), imgOri.type());

	cornerHarris(imgGray, imgHarris, blockSize, apertureSize, k);

	Mat imgDst, imgDstScale;
	imgDst = Mat::zeros(imgHarris.size(), imgHarris.type());
	normalize(imgHarris, imgDst, 0, 255, NORM_MINMAX);

	convertScaleAbs(imgDst, imgDstScale);

	for (int row = 0; row < imgDstScale.rows; row++)
	{
		for (int col = 0; col < imgDstScale.cols; col++)
		{
			int rsp = (int)imgDstScale.at<uchar>(row, col);
			if (rsp > 85)
			{
				int b = rng.uniform(0, 256);
				int g = rng.uniform(0, 256);
				int r = rng.uniform(0, 256);
				circle(imgOri, Point(col, row), 5, Scalar(b, g, r), 2);
			}
		}
	}

	imshow("imgHarris", imgOri);

	return true;
}

static bool shiTomas(Mat imgOri)
{
	int maxCorners = 500;
	double quality_level = 0.01;
	double minDistance = 0.04;

	Mat imgGray, imgShiTomas;
	cvtColor(imgOri, imgGray, COLOR_BGR2GRAY);

	vector<Point2f> corners;

	goodFeaturesToTrack(imgGray, corners, maxCorners, quality_level, minDistance, Mat(), 3, false);

	for (int i = 0; i < corners.size(); i++) {
		int b = rng.uniform(0, 256);
		int g = rng.uniform(0, 256);
		int r = rng.uniform(0, 256);
		circle(imgOri, corners[i], 5, Scalar(b, g, r), 3, 8, 0);
	}

	imshow("imgShiTomas", imgOri);

	return true;
}